java/src/org/openqa/selenium/grid/node/local/LocalNode.java

```diff
@@ -805,20 +805,33 @@ private Session createExternalSession(
     }
 
     // Check if the user wants to use BiDi
-    boolean webSocketUrl = toUse.asMap().containsKey("webSocketUrl");
-    // Add se:bidi if necessary to send the bidi url back
-    boolean bidiSupported = isSupportingBiDi || toUse.getCapability("se:bidi") != null;
-    if (bidiSupported && bidiEnabled && webSocketUrl) {
-      String bidiPath = String.format("/session/%s/se/bidi", other.getId());
-      toUse = new PersistentCapabilities(toUse).setCapability("se:bidi", rewrite(bidiPath));
+    // This will be null if the user has not set the capability.
+    Object webSocketUrl = toUse.getCapability("webSocketUrl");
+
+    // In case of Firefox versions that do not support webSocketUrl, it returns the capability as it
+    // is i.e. boolean value. So need to check if it is a string.
+    // Check if the Node supports BiDi and if the client wants to use BiDi.
+    boolean bidiSupported = isSupportingBiDi && (webSocketUrl instanceof String);
+    if (bidiSupported && bidiEnabled) {
+      String biDiUrl = (String) other.getCapabilities().getCapability("webSocketUrl");
+      URI uri = null;
+      try {
+        uri = new URI(biDiUrl);
+      } catch (URISyntaxException e) {
+        throw new IllegalArgumentException("Unable to create URI from " + uri);
+      }
+      toUse =
+          new PersistentCapabilities(toUse)
+              .setCapability("se:gridWebSocketUrl", uri)
+              .setCapability("webSocketUrl", rewrite(uri.getPath()));
```

Comment:
Why don't we keep using the `se/bidi` endpoint? Then the matching in `ProxyWebNodeSockets` is easier. We will still need to read the `webSocketUrl` in `DriverServiceSessionFactory` and get the internal one, but at least the client code does not need to know anything about it.

Reply:
The matching logic has not changed in `ProxyWebNodeSockets`. It is now actually following the pattern of the WebSocket URL i.e.` /session/{sessionId}`. Do you mean it will impact performance while matching?

Reply:
I only meant it from the simplicity point of view, as every functionality has its own endpoint. So, leaving it the way it was:
```
private static final UrlTemplate BIDI_TEMPLATE = new UrlTemplate("/session/{sessionId}/se/bidi");
```